pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,349 @@
+		}
+
+		currentLEUserHash := currentLEUser.toHash()
+
+		// Domain, LE environment or LE email changed -> delete secret for re-creation
+		if currentLEUserHash != newLEUserHash {
+			logrus.Infof("deleting let's encrypt secret due to config change: %v -> %v", currentLEUser, newLEUser)
+			err = client.Delete(ctx, leAccountSecret)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return currentLEUser, nil
+		}
+	}
+
+	if err := newLEUser.Register(); err != nil {
+		return nil, fmt.Errorf("problem registering Let's Encrypt User: %w", err)
+	}
+
+	sec, err := newLEUser.ToSecret()
+	if err != nil {
+		return nil, fmt.Errorf("problem creating Let's Encrypt User secret: %w", err)
+	}
+
+	if err := client.Create(ctx, sec); err != nil {
+		return nil, fmt.Errorf("problem creating Let's Encrypt User secret: %w", err)
+	}
+
+	logrus.Infof("Registered Let's Encrypt User: %s", newLEUser.Name)
+
+	return newLEUser, nil
+
+}
+
+func (u *LEUser) EnsureWildcardCertificateSecret(ctx context.Context, client kclient.Client, dnsendpoint, domain, token string) (*corev1.Secret, error) {
+
+	sec := &corev1.Secret{}
+	secErr := client.Get(ctx, router.Key(system.Namespace, system.TLSSecretName), sec)
+	if secErr != nil && !apierrors.IsNotFound(secErr) {
+		// error fetching the existing secret, but it could exist
+		return nil, secErr
+	}
+
+	leSettingsHash := u.toHash()
+
+	// Existing LE Cert Secret found, let's check if it's still valid
+	if secErr == nil {
+
+		// (a) domain or let's encrypt user settings changed -> renew
+		if sec.Annotations[labels.AcornLetsEncryptSettingsHash] != leSettingsHash {
+			logrus.Info("domain or let's encrypt settings changed, renewing wildcard certificate")
+		} else {
+
+			x509crt, err := certcrypto.ParsePEMCertificate([]byte(sec.Data[corev1.TLSCertKey]))
+			if err != nil {
+				// (b) unreadable certificate -> renew
+				logrus.Errorf("problem parsing existing TLS secret: %v", err)
+			} else {
+				timeToExpire := x509crt.NotAfter.Sub(time.Now().UTC())
+				if timeToExpire > 7*24*time.Hour {
+					// (c) cert is still valid for more than 7 days -> return it
+					logrus.Infof("existing TLS secret %s is still valid until %s (%d hours)", x509crt.Subject.CommonName, x509crt.NotAfter, int(timeToExpire.Hours()))
+					return sec, nil
+				} else {
+					// (d) cert is expired -> renew
+					logrus.Infof("existing TLS secret %s is expiring after %s (%d hours), renewing it...", x509crt.Subject.CommonName, x509crt.NotAfter, int(timeToExpire.Hours()))
+				}
+			}
+		}
+	}
+
+	cert, err := u.GenerateWildcardCert(dnsendpoint, domain, token)
+	if err != nil {
+		return nil, fmt.Errorf("problem generating wildcard certificate: %w", err)
+	}
+
+	x509crt, err := certcrypto.ParsePEMCertificate([]byte(cert.Certificate))
+
+	sec = &corev1.Secret{
+		Type: corev1.SecretTypeTLS,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.TLSSecretName,
+			Namespace: system.Namespace,
+			Annotations: map[string]string{
```

Comment:
Of these annotations, only the hash is ever read. What is the intention for the other three? Just informational?

Reply:
They make debugging easier in case anything goes wrong, but I'm totally fine with removing them as they're only informational.